File: app/src/main/java/com/qust/assistant/widget/lesson/LessonRender.java
package com.qust.assistant.widget.lesson;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.Build;
import android.util.TypedValue;

import androidx.annotation.NonNull;

import com.qust.assistant.model.LessonTableViewModel;
import com.qust.assistant.model.lesson.Lesson;
import com.qust.assistant.model.lesson.LessonGroup;
import com.qust.assistant.util.ColorUtil;
import com.qust.assistant.util.DateUtil;
import com.qust.assistant.util.ParamUtil;
import com.qust.assistant.util.SettingUtil;

import java.text.ParseException;
import java.util.Calendar;
import java.util.Date;

/**
 * 总课表界面的渲染器
 */
public class LessonRender{
	
	private static final String[] WEEK_STRING = {"周一", "周二", "周三", "周四", "周五", "周六", "周日"};
	
	/**
	 * 课程间距
	 */
	private static final int LESSON_PADDING = 3;
	
	/**
	 * 开学日期
	 */
	private Calendar startDate;
	
	/**
	 * 今天的日期
	 */
	private Calendar currentDay;
	
	/**
	 * 总周数
	 */
	private int totalWeeks;
	
	/**
	 * 显示全部课程
	 */
	private boolean showAllLesson;
	
	/**
	 * 隐藏已结课程
	 */
	private boolean hideFinishLesson;
	
	/**
	 * 隐藏教师
	 */
	private boolean hideTeacher;
	
	private int baseLine;
	
	private int textHeight;
	
	/**
	 * 左侧时间和顶部日期的宽度
	 */
	public int timeWidth, dateHeight;
	
	/**
	 * 最小的一节课的大小
	 */
	private int cellWidth, cellHeight;
	
	/**
	 * 不同文本之间的间距
	 */
	private int linePadding;
	
	private Paint paint, paintT;
	
	private LessonCell[][][] lessons;
	
	
	public LessonRender(@NonNull Context context){
		
		startDate = Calendar.getInstance();
		
		currentDay = Calendar.getInstance();
		
		totalWeeks = LessonTableViewModel.getTotalWeek();
		
		paint = new Paint(Paint.FILTER_BITMAP_FLAG);
		paint.setStyle(Paint.Style.FILL);
		paint.setAntiAlias(true);
		paint.setStrokeWidth(3);
		
		paintT = new Paint();
		paintT.setDither(true);
		paintT.setAntiAlias(true);
		paintT.setSubpixelText(true);
		paintT.setTextAlign(Paint.Align.CENTER);
		
		paintT.setTextSize(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 12, context.getResources().getDisplayMetrics()));
		
		baseLine = (int)(paintT.getTextSize() / 2 + (paintT.getFontMetrics().descent - paintT.getFontMetrics().ascent) / 2 - paintT.getFontMetrics().descent);
		
		textHeight = (int)(paintT.getTextSize() + 3);
		
		timeWidth = ParamUtil.dp2px(context, 48);
		linePadding = ParamUtil.dp2px(context, 4);
		
		dateHeight = 40 + textHeight * 2;
		
		hideTeacher = SettingUtil.getBoolean(SettingUtil.KEY_HIDE_TEACHER, false);
		showAllLesson = SettingUtil.getBoolean(SettingUtil.KEY_SHOW_ALL_LESSON, false);
		hideFinishLesson = SettingUtil.getBoolean(SettingUtil.KEY_HIDE_FINISH_LESSON, false);
	}
	
	
	/**
	 * 设置学期信息
	 * @param _startDate 开学时间
	 * @param _totalWeeks 总周数
	 */
	public void setTermData(String _startDate, int _totalWeeks){
		totalWeeks = _totalWeeks;
		
		try{
			Date date = DateUtil.YMD.parse(_startDate);
			if(date != null) startDate.setTime(date);
		}catch(ParseException ignored){
		}
	}
	
	/**
	 * 设置 / 更新 课表信息，需要在 {@link LessonRender#setMeasureData} 之后调用
	 * @param lessonGroups 课表信息
	 */
	public void setLessonData(@NonNull LessonGroup[][] lessonGroups){
		lessons = new LessonCell[totalWeeks][lessonGroups.length][lessonGroups[0].length];
		
		for(int week = 0; week < totalWeeks; week++){
			for(int i = 0; i < lessonGroups.length; i++){
				int j = 0;
				while(j < lessonGroups[0].length){
					LessonGroup lessonGroup = lessonGroups[i][j];
					if(lessonGroup != null){
						Lesson lesson = lessonGroup.getCurrentLesson(week + 1);
						if(lesson != null){
							lessons[week][i][j] = new LessonCell(lesson.color, lesson);
							j += lesson.len;
							continue;
						}
					}
					j++;
				}
			}
		}
		
		if(showAllLesson){
			for(int week = 0; week < totalWeeks; week++){
				for(int i = 0; i < lessonGroups.length; i++){
					int j = 0;
					while(j < lessonGroups[0].length){
						LessonCell lessonCell = lessons[week][i][j];
						if(lessonCell != null){
							j += lessonCell.len;
							continue;
						}
						LessonGroup lessonGroup = lessonGroups[i][j];
						if(lessonGroup != null){
							// 本周该时间无课但是其他周有课
							Lesson lesson = lessonGroup.findLesson(week + 1, !hideFinishLesson);
							if(lesson != null){
								boolean conflictFlag = false;
								for(int p = j + 1, l = 0; p < lessonGroups[0].length && l < lesson.len - 1; p++, l++){
									// 向下检查是否有冲突
									if(lessons[week][i][p] != null){
										conflictFlag = true;
										break;
									}
								}
								if(!conflictFlag){
									lessons[week][i][j] = new LessonCell(-1, lesson);
									j += lesson.len;
									continue;
								}
							}
						}
						j++;
					}
				}
			}
		}
	}
	
	/**
	 * 设置 View Measure 数据
	 */
	public void setMeasureData(int measuredWidth, int measuredHeight){
		cellWidth = (measuredWidth - timeWidth) / WEEK_STRING.length;
		cellHeight = (measuredHeight - dateHeight) / LessonTableViewModel.getLessonTimeText()[0].length;
	}
	
	/**
	 * 获取点击位置的课程
	 * @return [week, count, 是否有课(0为无课)]
	 */
	public int[] getClickLesson(int showWeek, int downX, int downY){
		if(downX < timeWidth || downY < dateHeight) return new int[] { -1, -1 , 0 };
		
		int y = downY - dateHeight;
		
		// 计算点击的位置在第几周
		int currentWeek = (downX - timeWidth) / cellWidth;
		
		for(int i = 0; i < lessons[showWeek][currentWeek].length; i++){
			
			LessonCell lesson = lessons[showWeek][currentWeek][i];
			
			if(lesson != null && y < lesson.len * cellHeight){
				return new int[] { currentWeek, i , 1 };
			}
			
			if(y < cellHeight){
				return new int[] { currentWeek, i , 0 };
			}
			
			y -= cellHeight;
		}
		
		return new int[] { -1, -1, 0 };
	}
	
	/**
	 * 绘制View
	 * @param week 绘制第几周，从0开始
	 */
	public void drawView(Canvas canvas, int week){
		drawTime(canvas);
		drawDate(canvas, week);
		drawLessons(canvas, week);
	}
	
	/**
	 * 绘制选中高亮框
	 * @param week 星期几
	 * @param count 第几节
	 * @param len 课程长度
	 */
	public void drawHighlightBox(Canvas canvas, int week, int count, int len){
		paint.setStyle(Paint.Style.STROKE);
		paint.setColor(Color.rgb(0, 176, 255));
		if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP){
			canvas.drawRoundRect(week * cellWidth + timeWidth + LESSON_PADDING, count * cellHeight + dateHeight + LESSON_PADDING, week * cellWidth + cellWidth + timeWidth - LESSON_PADDING, count * cellHeight + cellHeight * len + dateHeight - LESSON_PADDING, 16, 16, paint);
		}else{
			canvas.drawRoundRect(new RectF(week * cellWidth + timeWidth + LESSON_PADDING, count * cellHeight + dateHeight + LESSON_PADDING, week * cellWidth + cellWidth + timeWidth - LESSON_PADDING, count * cellHeight + cellHeight * len + dateHeight - LESSON_PADDING), 16, 16, paint);
		}
		paint.setStyle(Paint.Style.FILL);
	}
	
	
	protected void drawTime(Canvas canvas){
		paintT.setColor(Color.GRAY);
		String[][] timeText = LessonTableViewModel.getLessonTimeText();
		int x = timeWidth / 2;
		int y = dateHeight + baseLine + (cellHeight - textHeight * 2) / 2;
		for(int i = 0; i < lessons[0][0].length; i++){
			canvas.drawText(timeText[0][i], x, y, paintT);
			canvas.drawText(timeText[1][i], x, y + textHeight, paintT);
			y += cellHeight;
		}
	}
	
	protected void drawDate(Canvas canvas, int week){
		
		Calendar c = (Calendar) startDate.clone();
		
		c.add(Calendar.WEEK_OF_YEAR, week);
		int dayOfWeek = c.get(Calendar.DAY_OF_WEEK) - 2;
		c.add(Calendar.DATE, -dayOfWeek);
		
		int y = 20 + baseLine;
		
		for(int i = 0; i < WEEK_STRING.length; i++){
			
			if(currentDay.get(Calendar.DATE) == c.get(Calendar.DATE) && currentDay.get(Calendar.MONTH) == c.get(Calendar.MONTH)){
				paintT.setColor(ColorUtil.TEXT_COLORS[0]);
			}else{
				paintT.setColor(Color.GRAY);
			}
			
			String day = DateUtil.MD.format(c.getTime());
			
			canvas.drawText(WEEK_STRING[i], timeWidth + cellWidth / 2 + i * cellWidth, y, paintT);
			canvas.drawText(day, timeWidth + cellWidth / 2 + i * cellWidth, y + textHeight, paintT);
			
			c.add(Calendar.DATE, 1);
		}
	}
	
	protected void drawLessons(Canvas canvas, int showWeek){
		int x = timeWidth;
		
		for(int i = 0; i < lessons[showWeek].length; i++, x += cellWidth){
			
			int y = dateHeight;
			
			for(int j = 0; j < lessons[showWeek][0].length; j++, y += cellHeight){
				
				LessonCell lesson = lessons[showWeek][i][j];
				
				if(lesson == null) continue;
				
				paint.setColor(lesson.color == -1 ? ColorUtil.BACKGROUND_COLOR_SECOND : ColorUtil.BACKGROUND_COLORS[lesson.color]);
				paintT.setColor(lesson.color == -1 ? ColorUtil.TEXT_COLOR_SECOND : ColorUtil.TEXT_COLORS[lesson.color]);
				
				if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP){
					canvas.drawRoundRect(x + LESSON_PADDING, y + LESSON_PADDING, x + cellWidth - LESSON_PADDING, y + cellHeight * lesson.len - LESSON_PADDING, 16, 16, paint);
				}else{
					canvas.drawRoundRect(new RectF(x + LESSON_PADDING, y + LESSON_PADDING, x + cellWidth - LESSON_PADDING, y + cellHeight * lesson.len - LESSON_PADDING), 16, 16, paint);
				}
				
				int lineY = y + baseLine + (cellHeight * lesson.len - textHeight * lesson.lines - linePadding * (hideTeacher ? 1 : 2)) / 2;
				
				int c = lesson.lines + (hideTeacher ? 1 : 2);
				for(int n = 0; n < c; n++){
					if(lesson.data[n] == null){
						lineY += linePadding;
					}else {
						canvas.drawText(lesson.data[n], x + cellWidth / 2, lineY, paintT);
						lineY += textHeight;
					}
				}
			}
		}
	}

	/**
	 * 字符串分行
	 * @param src 要分行的字符串
	 * @param des 分行后的字符串
	 * @param maxWidth 一行最大的长度
	 * @param desPos 目标数组开始的位置
	 * @param maxLine 最大支持的行数
	 */
	private int splitString(@NonNull String src, String[] des, int maxWidth, int desPos, int maxLine){
		int srcPos = 0;
		int lines = 0;
		int length = src.length();
		for(int i = srcPos; i < length;){
			i = srcPos + paintT.breakText(src, i, length, true, maxWidth, null);
			des[desPos + lines] = src.substring(srcPos, i);
			srcPos = i;
			if(lines++ == maxLine){
				lines--;
				return lines;
			}
		}
		return lines;
	}
	
	
	private class LessonCell{
		/**
		 * 课程长度
		 */
		public int len;
		
		/**
		 * 课程颜色
		 */
		public int color;
		
		/**
		 * 文本信息的行数
		 */
		public int lines;
		
		/**
		 * 课程文本信息
		 */
		public String[] data;
		
		public LessonCell(int color, @NonNull Lesson lesson){
			this.color = color;
			this.len = lesson.len;

			data = new String[len * 3 + (hideTeacher ? 1 : 2)];
			
			lines = splitString(lesson.name, data, cellWidth - (LESSON_PADDING << 2), 0, len * 3 - 2);
			lines += splitString(lesson.place, data, cellWidth - (LESSON_PADDING << 2), lines + 1, len * 3 - lines - 1);
			if(!hideTeacher) lines += splitString(lesson.teacher, data, cellWidth - (LESSON_PADDING << 2), lines + 2, len * 3 - lines);
		}
	}
	
}

File: composeApp/src/androidMain/kotlin/com/qust/helper/ui/widget/LessonRender.kt
package com.qust.helper.ui.widget

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import androidx.compose.runtime.MutableIntState
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.qust.helper.data.Data
import com.qust.helper.data.lesson.Lesson
import com.qust.helper.data.lesson.LessonTable
import com.qust.helper.model.LessonTableRepository
import com.qust.helper.ui.theme.BACKGROUND_COLORS
import com.qust.helper.ui.theme.BACKGROUND_COLOR_SECOND
import com.qust.helper.ui.theme.TEXT_COLORS
import com.qust.helper.ui.theme.TEXT_COLOR_SECOND
import java.util.Date

/**
 * 总课表界面的渲染器
 */
class LessonRender(
	startDay: MutableState<Date> = LessonTableRepository._startDay,
	totalWeek: MutableIntState = LessonTableRepository._totalWeek,
	lessonTable: MutableState<LessonTable> = LessonTableRepository._lessonTable,
) {

	private var baseLine = 0
	private var textHeight = 0

	/**
	 * 左侧时间和顶部日期的宽度
	 */
	private var timeWidth = 0
	private var dateHeight = 0

	/**
	 * 最小的一节课的大小
	 */
	private var cellWidth = 0
	private var cellHeight = 0

	/**
	 * 不同文本之间的间距
	 */
	private var linePadding = 0

	private var composeWidth: Int = 0
	private var composeHeight: Int = 0

	private val paint: Paint = Paint(Paint.FILTER_BITMAP_FLAG).also { paint ->
		paint.style = Paint.Style.FILL
		paint.isAntiAlias = true
		paint.strokeWidth = 3f
	}

	private val paintT: Paint = Paint().also { paintT ->
		paintT.isDither = true
		paintT.isAntiAlias = true
		paintT.isSubpixelText = true
		paintT.textAlign = Paint.Align.CENTER
	}

	private val hideTeacher by LessonTableRepository._hideTeacher
	private val showAllLesson by LessonTableRepository._showAllLesson
	private val hideFinishLesson by LessonTableRepository._hideFinishLesson

	private var _startDay = startDay
	val startDay: Date
		get() = _startDay.value

	private var _totalWeek = totalWeek
	val totalWeek: Int
		get() = _totalWeek.intValue

	private var _lessonTable = lessonTable
	val lessonTable: LessonTable
		get() = _lessonTable.value

	private var lessonRenderData by mutableStateOf(LessonRenderData(hideTeacher, paintT))

	val lockLesson by LessonTableRepository._lockLesson

	/**
	 * 设置 View Measure 数据
	 * 必须调用，不然无法显示
	 */
	fun setMeasureData(measuredWidth: Int, measuredHeight: Int, density: Density) {
		if(measuredWidth == composeWidth && measuredHeight == composeHeight) return
		with(density) {
			linePadding = 4.dp.toPx().toInt()
			paintT.textSize = 12.sp.toPx()
			baseLine = (paintT.textSize / 2 + (paintT.fontMetrics.descent - paintT.fontMetrics.ascent) / 2 - paintT.fontMetrics.descent).toInt()
			textHeight = (paintT.textSize + 3).toInt()
		}
		composeWidth = measuredWidth
		composeHeight = measuredHeight
		cellWidth = (measuredWidth - timeWidth) / 7
		cellHeight = (measuredHeight - dateHeight) / 10
		lessonRenderData = LessonRenderData(hideTeacher, paintT, (cellWidth - (LESSON_PADDING shl 2)), lessonTable)
		lessonRenderData.calcLessonData()
	}

	fun updateLessonTable() {
		if(cellWidth != 0 && cellHeight != 0) {
			lessonRenderData = LessonRenderData(hideTeacher, paintT, (cellWidth - (LESSON_PADDING shl 2)), lessonTable)
			lessonRenderData.calcLessonData()
		}
	}

	/**
	 * 获取点击位置的课程
	 * @param week 当前周(从0开始)
	 * @param downX 点击X坐标
	 * @param downY 点击Y坐标
	 */
	fun getClickLesson(week: Int, downX: Int, downY: Int): SelectLesson {
		if(downX < timeWidth || downY < dateHeight) {
			return SelectLesson(-1, -1, null)
		}

		// 计算点击的位置是星期几
		val dayOfWeek = (downX - timeWidth) / cellWidth
		if(dayOfWeek >= Data.WEEK_STRING.size) {
			return SelectLesson(-1, -1, null)
		}
		var y = downY - dateHeight

		for(timeSlot in lessonRenderData.lessons[dayOfWeek].indices) {
			lessonRenderData.lessons[dayOfWeek][timeSlot]?.let{ holder ->
				var lessonData = holder.current(week)
				if(lessonData == null && showAllLesson) {
					lessonData = holder.findLesson(week, !hideFinishLesson)
				}
				if(lessonData != null) {
					if(y < lessonData.len * cellHeight) {
						return SelectLesson(dayOfWeek, timeSlot, lessonRenderData.getLessonByHolder(week, dayOfWeek, timeSlot))
					}
				}
			}

			if(y < cellHeight) {
				return SelectLesson(dayOfWeek, timeSlot, null)
			}
			y -= cellHeight
		}
		return SelectLesson(-1, -1, null)
	}

	fun hasNextLesson(week: Int, dayOfWeek: Int, timeSlot: Int): Boolean {
		return lessonRenderData.lessons[dayOfWeek][timeSlot]?.hasNext(week) ?: false
	}

	fun nextLesson(week: Int, dayOfWeek: Int, timeSlot: Int): Lesson {
		lessonRenderData.lessons[dayOfWeek][timeSlot]?.next(week)
		return lessonRenderData.getLessonByHolder(week, dayOfWeek, timeSlot)
	}

	/**
	 * 绘制View
	 * @param week 绘制第几周，从0开始
	 */
	fun drawView(canvas: Canvas, week: Int) {
		var x = timeWidth
		for(i in lessonRenderData.lessons.indices) {
			var y = dateHeight
			for(j in lessonRenderData.lessons[0].indices) {

				val holder: LessonRenderData.LessonHolder? = lessonRenderData.lessons[i][j]
				if(holder == null) { y += cellHeight; continue }

				var lesson = holder.current(week)
				if(lesson == null) {
					if(!showAllLesson) { y += cellHeight; continue }
					lesson = holder.findLesson(week, !hideFinishLesson)
					if(lesson == null) { y += cellHeight; continue }

					paint.color = BACKGROUND_COLOR_SECOND
					paintT.color = TEXT_COLOR_SECOND
				} else {
					paint.color = BACKGROUND_COLORS[lesson.color]
					paintT.color = TEXT_COLORS[lesson.color]
				}

				canvas.drawRoundRect((x + LESSON_PADDING).toFloat(), (y + LESSON_PADDING).toFloat(), (x + cellWidth - LESSON_PADDING).toFloat(), (y + cellHeight * lesson.len - LESSON_PADDING).toFloat(), 16f, 16f, paint)
				canvas.drawText(if(lesson.type == 0) "A" else "U", (x + (LESSON_PADDING shl 2) + 3).toFloat(), (y + baseLine + (LESSON_PADDING shl 2)).toFloat(), paintT)

				if(holder.lessonCount[week] > 1) {
					canvas.drawText((holder.index[week] + 1).toString() + "/" + holder.lessonCount[week], (x + cellWidth / 2).toFloat(), (y + cellHeight * lesson.len - textHeight + baseLine - (LESSON_PADDING shl 2)).toFloat(), paintT)
				}

				var lineY: Int = y + baseLine + (cellHeight * lesson.len - textHeight * lesson.lines - linePadding * if(hideTeacher) 1 else 2) / 2
				for(n in lesson.data.indices) {
					lineY += if(lesson.data[n] == null) {
						linePadding
					} else {
						canvas.drawText(lesson.data[n]!!, (x + cellWidth / 2).toFloat(), lineY.toFloat(), paintT)
						textHeight
					}
				}
				y += cellHeight
			}
			x += cellWidth
		}
	}

	/**
	 * 绘制选中高亮框
	 * @param week 星期几
	 * @param count 第几节
	 * @param len 课程长度
	 */
	fun drawHighlightBox(canvas: Canvas, week: Int, count: Int, len: Int) {
		paint.style = Paint.Style.STROKE
		paint.color = Color.rgb(0, 176, 255)
		canvas.drawRoundRect(
			(week * cellWidth + timeWidth + LESSON_PADDING).toFloat(),
			(count * cellHeight + dateHeight + LESSON_PADDING).toFloat(),
			(week * cellWidth + cellWidth + timeWidth - LESSON_PADDING).toFloat(),
			(count * cellHeight + cellHeight * len + dateHeight - LESSON_PADDING).toFloat(),
			16f,
			16f,
			paint
		)
		paint.style = Paint.Style.FILL
	}


	data class SelectLesson(
		val dayOfWeek: Int,
		val timeSlot: Int,
		val lesson: Lesson?
	)


	companion object {
		/**
		 * 课程间距
		 */
		private const val LESSON_PADDING = 3
	}
}
